File: SubXor.py
from collections import defaultdict
from types import GeneratorType


def bootstrap(f, stack=[]):
    def wrappedfunc(*args, **kwargs):
        if stack:
            return f(*args, **kwargs)
        else:
            to = f(*args, **kwargs)
            while True:
                if type(to) is GeneratorType:
                    stack.append(to)
                    to = next(to)
                else:
                    stack.pop()
                    if not stack:
                        break
                    to = stack[-1].send(to)
            return to

    return wrappedfunc


class BinaryTrie:
    def __init__(self, max_bits):
        self.max_bits = max_bits
        self.trie = [[0, 0] for _ in range(10**5 * max_bits)]
        self.count = [0] * (10**5 * max_bits)
        self.size = 0

    @bootstrap
    def _insert(self, node, mbits, num):
        if mbits < 0:
            yield None
            return
        self.count[node] += 1
        bit = (num >> mbits) & 1
        if not self.trie[node][bit]:
            self.size += 1
            self.trie[node][bit] = self.size
        yield self._insert(self.trie[node][bit], mbits - 1, num)
        yield None

    def insert(self, num):
        self._insert(0, self.max_bits, num)

    @bootstrap
    def _query(self, node, mbits, num):
        if mbits < 0:
            yield 0
            return
        bit = (num >> mbits) & 1
        target = bit
        max_xor = 0
        if self.trie[node][1 - bit]:
            target = 1 - bit
        ret = yield self._query(self.trie[node][target], mbits - 1, num)
        max_xor = ret | (target << mbits)
        yield max_xor

    def query_max_xor(self, num):
        """返回 num 与 Trie 中已有数字异或的最大值"""
        return self._query(0, self.max_bits, num)

    @bootstrap
    def _query_under(self, node, mbits, num, K):
        if mbits < 0:
            yield 0
            return
        bit_N = (num >> mbits) & 1
        bit_K = (K >> mbits) & 1
        ans = 0
        same = self.trie[node][bit_N]
        diff = self.trie[node][1 - bit_N]
        if bit_K == 1:
            if same:
                ans += self.count[same]
            if diff:
                ans += yield self._query_under(diff, mbits - 1, num, K)
        else:
            if same:
                ans += yield self._query_under(same, mbits - 1, num, K)

        yield ans

    def query_under(self, num, K):
        return self._query_under(0, self.max_bits, num, K)


t = int(input())
n, k = map(int, input().split())
a = [*map(int, input().split())]
pre = [0] * (n + 1)
for i in range(1, n + 1):
    pre[i] = pre[i - 1] ^ a[i - 1]
ans = 0
bt = BinaryTrie(20)
bt.insert(pre[0])
for i in range(1, n + 1):
    ans += bt.query_under(pre[i], k)
    bt.insert(pre[i])
print(ans)

File: SubXor.kt
class BinaryTrie(val maxBits: Int, maxNodes: Int) {
    // 模拟 Python 的 trie = [[0, 0]]
    val trie = Array(maxNodes + 1) { IntArray(2) }
    val count = IntArray(maxNodes + 1)
    var size = 0

    // 插入入口
    fun insert(num: Int) {
        _insert(0, maxBits, num)
    }

    // 递归插入逻辑
    private fun _insert(node: Int, mbits: Int, num: Int) {
        // 路径经过即增加计数
        count[node]++

        if (mbits < 0) return

        val bit = (num ushr mbits) and 1
        
        if (trie[node][bit] == 0) {
            size++
            // 这里的 trie[size] 已经在初始化时准备好了
            trie[node][bit] = size
        }
        
        _insert(trie[node][bit], mbits - 1, num)
    }

    // 查询最大异或值入口
    fun queryMaxXor(num: Int): Int {
        return _queryMaxXor(0, maxBits, num)
    }

    // 递归查询最大异或值逻辑
    private fun _queryMaxXor(node: Int, mbits: Int, num: Int): Int {
        if (mbits < 0) return 0

        val bit = (num ushr mbits) and 1
        val target = 1 - bit

        return if (trie[node][target] != 0) {
            // 能够走相反位：该位异或结果为 1，加上低位的递归结果
            (1 shl mbits) or _queryMaxXor(trie[node][target], mbits - 1, num)
        } else {
            // 只能走相同位：该位异或结果为 0
            _queryMaxXor(trie[node][bit], mbits - 1, num)
        }
    }
}

fun main() {
    val bt = BinaryTrie(30, 100000)
    
    bt.insert(123)
    bt.insert(243)
    
    println("Max XOR with 20: ${bt.queryMaxXor(20)}")
    // 调试打印根节点的 count (应该是 2)
    println("Root node count: ${bt.count[0]}")
}